### ISSUE TITLE
Mismatched conversion of braces in title on export means field never gets closed

### ISSUE THREAD

**Issue by robbyblum** [retorquere/zotero-better-bibtex#1218]
**Report ID:** LJMFG4C2-euc

**Exporter used:** Bibtex exporter

**Expected behavior:** something like `title = {Resolving biomolecular motion and interactions by $R_2$ and $R_{1\text{ρ}}$ relaxation dispersion {{NMR}}},`

**Actual behavior:**   `title = {Resolving Biomolecular Motion and Interactions by \${{R}}\_2\$ and \${{R}}\_\{1\textbackslash{}text\{{$\rho\rbrace\rbrace\$$} Relaxation Dispersion {{NMR}}},`

The title in the Zotero entry is `Resolving biomolecular motion and interactions by $R_2$ and $R_{1\text{ρ}}$ relaxation dispersion NMR`.

I think that this exported more "correctly" on a previous version of the translator, but under the most recent version, it's trying to do a bunch of crazy conversions and it looks to me like it ends up missing a } (by converting it into \rbrace i guess?), which means the title field never gets closed, and Bibtex gets upset and ignores this field and the entire rest of the .bib file.

If I change the Zotero title to `Resolving biomolecular motion and interactions by $R_2$ and $R_{1\rho}$ relaxation dispersion NMR`, then BBT converts it to `  title = {Resolving Biomolecular Motion and Interactions by \${{R}}\_2\$ and \${{R}}\_\{1\textbackslash{}rho\}\$ Relaxation Dispersion {{NMR}}},`. This is still *wrong*, because it's escaping out all the LaTeX commands, but it doesn't have the mismatched braces issue. I guess the unicode rho is throwing it off somewhere.

In a sense, I guess this is probably because I was using Zotero "incorrectly" by having LaTeX commands in the paper title, but this was the way I had previously managed to get it to format this title correctly due to some annoying font I had to use on a previous project. I'm not sure what the "correct" way to keep these sort of LaTeX commands in a Zotero field is, if there is one.

**Comment by robbyblum** [comment]
Of course, I figure out "the "correct" way to keep these sort of LaTeX commands in a Zotero field is" immediately after I post this (using `<pre> ... </pre>` or the #LaTeX tag). I still think this is a bug, though, because the converter shouldn't be screwing up its own braces...

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.1.99.4277](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.99.4277.xpi) ("fixes #1218")

Install in Zotero by downloading [test build 5.1.99.4277](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.1.99.4277.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
give 4277 a go

**Comment by retorquere** [comment]
Also: https://retorque.re/zotero-better-bibtex/exporting/scripting/#detect-and-protect-latex-math-formulas (but I'd really appreciate it if you could confirm 4277 fixes it for you -- I've added the sample from LJMFG4C2 as a testcase, so it should, but confirmation is always nice)

**Comment by robbyblum** [comment]
Looks like 4277 fixes it for me:
```
title = {Resolving Biomolecular Motion and Interactions by \${{R}}\_2\$ and \${{R}}\_\{1\textbackslash{}text\{{$\rho\rbrace\rbrace\$$} Relaxation Dispersion \vphantom{\}\}}{{NMR}}},
```
Which is ugly ("Resolving biomolecular motion and interactions by $R 2$ and $R {1\text{ρ}}$ Relaxation Dispersion NMR.") but is what it's supposed to be doing! No more mysterious invalid-end-of-file errors from BibTeX.

**Comment by retorquere** [comment]
Awesome. Yeah, ugly; once Zotero gains mathjax capabilities (hopefully someday) I could properly export math sections. Until then, the postscript could help, and of course `<pre>` if you want to force it. The postscript really just inserts `<pre>` tags when it finds a `$`, and it isn't sophisticated; any pair of `$` signs gets the same treatment.

**Comment by retorquere** [comment]
5.1.100 is building and should be out in 10 minutes or so (it's running the full test suite again), but it's just a rebadged 4277, and 4277 includes the fixes from #1217, so you can just stick with what you have. Anywhere between now and two days, Zotero will check for updates and you'll get .100. You can also force a refresh from the addons screen if you want it sooner.

Thanks for reporting, and the confirmation on both.

**Comment by retorquere** [comment]
The `\rbrace` vs `\}` thing is because BBT is in [conservative](https://retorque.re/zotero-better-bibtex/installation/preferences/hidden-preferences/#mapunicode) mode BTW; by default, BBT seeks to minimize the number of `$`s generated, so when it is in text-mode, it aims to stay in text mode as long as possible, and once in math mode, it aims to stay in math mode as long as possible. If you have it prefer text (or math), it would have switched out of math mode to generate `\}` instead of `\rbrace`, you wouldn't have seen the missing brace, and it also wouldn't generate the balancing `vphantom`. 

The `vphantom` may not even be required any more; [the problem it aims to fix](https://tex.stackexchange.com/questions/230750/open-brace-in-bibtex-fields/230754#comment545453_230754) was reported and fixed 4 years ago, but the LaTeX world moves glacially slow, and I wouldn't put it past even large publishers to not have updated yet.

**Comment by robbyblum** [comment]
Thank you for the postscript; it should make my life a lot easier. One thing to put in the advanced options is much better than remembering to insert tags everywhere! The only odd thing I found with the postscript was that, for some reason, after trying it, the $R_2$ etc ended up typesetting as $r_2$. I just added braces around the Rs to fix that though. Probably just my bibliography style being too aggressive in forcing sentence case, or something.

Also, thanks for the tip about conservative mode! I couldn't fully figure out what issue the `vphantom` was supposed to fix, from that SX thread, but that's not my problem anyway. Thanks again for all the help!

**Comment by retorquere** [comment]
You can't win from label gun. Only I can close non-question issues, and label fun reopens issues if anyone comments so I don't forget them.

**Comment by retorquere** [comment]
WRT `$R_2$` etc ended up typesetting as `$r_2$`, try this postscript instead:

 ```
if (Translator.BetterTeX && reference.has.title) {
  reference.add({ name: 'title', value: item.title.replace(/(\$.*?\$)/g, '<script>{$1}</script>') });
}
```

(`<pre>` and `<script>` do the same thing, but I should have chosen `<script>` when I first implemented this rather than `<pre>`)

**Comment by retorquere** [comment]
> Also, thanks for the tip about conservative mode! I couldn't fully figure out what issue the vphantom was supposed to fix, from that SX thread, but that's not my problem anyway. Thanks again for all the help!

It's not really conservative mode itself that triggers it -- BBT always tries to balance out braces, and unbalanced braces is most likely to happen in conservative mode.

In `math` mode, `{` is changed to `\lbrace` and `}` to `\rbrace`; in `text` mode, `{` is changed to `\{` and `}` to `\}`, so that will usually balance out unless you have an actual `{`/`}` in your input.

In your case, the `{` was exported in `text` mode, BBT changed to `math` mode because of the `_`, and it was still in math mode when it came to `}`. The error fixed in #1218 was that I was testing the input for unbalancedness (counting `{` and `}`) rather than the output (counting `\{` and `\}`) and with mode switching -- badabing badaboom, unbalanced braces.
